segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,352 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+                    queueSize.incrementAndGet();
+                    SLTS_GC_SEGMENT_QUEUED.inc();
+                }, this.storageExecutor);
     }
 
     /**
-     * Adds given chunks to list of garbage chunks.
-     *
-     * @param chunksToDelete List of chunks to delete.
+     * Adds new chunk to track
+     * @param transactionId TransactionId
+     * @param chunktoTrack Name of chunk to track.
+     * @return A CompletableFuture that, when completed, will indicate the operation succeeded.
+     * If the operation failed, it will contain the cause of the failure.
      */
-    void addToGarbage(Collection<String> chunksToDelete) {
-        val currentTime = currentTimeSupplier.get();
+    CompletableFuture<Void> trackNewChunk(long transactionId, String chunktoTrack) {
+        Preconditions.checkState(null != taskQueue, "taskQueue must not be null.");
+        val startTime = currentTimeSupplier.get() + config.getGarbageCollectionDelay().toMillis();
+        // Simply add delete chunk task for newly tracked chunk and update metrics.
+        return taskQueue.addTask(taskQueueName, new TaskInfo(chunktoTrack, startTime, 0, TaskInfo.DELETE_CHUNK, transactionId))
+                .thenRunAsync(() -> {
+                    queueSize.incrementAndGet();
+                    SLTS_GC_CHUNK_NEW.inc();
+                }, this.storageExecutor);
+    }
 
-        chunksToDelete.forEach(chunkToDelete -> addToGarbage(chunkToDelete, currentTime + config.getGarbageCollectionDelay().toMillis(), 0));
+    /**
+     * Add the task to failed queue.
+     */
+    private CompletableFuture<Void> failTask(TaskInfo infoToRetire) {
+        Preconditions.checkState(null != taskQueue, "taskQueue must not be null.");
+        return taskQueue.addTask(failedQueueName, infoToRetire);
+    }
 
-        if (queueSize.get() >= config.getGarbageCollectionMaxQueueSize()) {
-            log.warn("{}: deleteGarbage - Queue full. Could not delete garbage. Chunks skipped", traceObjectId);
-        }
+    /**
+     * Perform delete segment related tasks.
+     */
+    private CompletableFuture<Void> deleteSegment(TaskInfo taskInfo) {
+        val streamSegmentName = taskInfo.getName();
+        ArrayList<String> chunksToDelete = new ArrayList<>();
+        @Cleanup
+        val txn = metadataStore.beginTransaction(false, streamSegmentName);
+        return txn.get(streamSegmentName)
+                .thenComposeAsync(storageMetadata -> {
+                    val segmentMetadata = (SegmentMetadata) storageMetadata;
+                    if (null == segmentMetadata) {
+                        log.debug("{}: deleteGarbage - Segment metadata does not exist. segment={}.", traceObjectId, streamSegmentName);
+                        return CompletableFuture.completedFuture(null);
+                    } else if (segmentMetadata.isActive()) {
+                        log.debug("{}: deleteGarbage - Segment is not marked as deleted. segment={}.", traceObjectId, streamSegmentName);
```

Comment:
Is that just `debug` that GC has been instructed to delete an active Segment? Looks like a wrong logic that may require us to notice, right? For this reason I don't know if `warn` may be needed here (unless this case is common and harmless for some reason that I ignore).

Reply:
This is okay.
In the new design we optimistically enqueue GC tasks **before**  SLTS transaction is committed. In case of transaction failure , the segment/chunks will be still marked active and that means transaction failed and GC task need to be fixed.
https://github.com/pravega/pravega/wiki/PDP-53-(Robust-Garbage-Collection-for-SLTS)#proposal